File: data/readable_format/cpp/file054.cpp
#include <iostream>
#include <cmath>

// Declare and define the DLAE2 function
void DLAE2(double A, double B, double C, double& RT1, double& RT2) {
    const double ONE = 1.0;
    const double TWO = 2.0;
    const double ZERO = 0.0;
    const double HALF = 0.5;
    double SM, DF, ADF, TB, AB, ACMN, ACMX, RT;

    SM = A + C;
    DF = A - C;
    ADF = std::abs(DF);
    TB = B + B;
    AB = std::abs(TB);

    if (std::abs(A) > std::abs(C)) {
        ACMX = A;
        ACMN = C;
    } else {
        ACMX = C;
        ACMN = A;
    }

    if (ADF > AB) {
        RT = ADF * std::sqrt(ONE + std::pow((AB / ADF), 2));
    } else if (ADF < AB) {
        RT = AB * std::sqrt(ONE + std::pow((ADF / AB), 2));
    } else {
        RT = AB * std::sqrt(TWO);
    }

    if (SM < ZERO) {
        RT1 = HALF * (SM - RT);
        RT2 = (ACMX / RT1) * ACMN - (B / RT1) * B;
    } else if (SM > ZERO) {
        RT1 = HALF * (SM + RT);
        RT2 = (ACMX / RT1) * ACMN - (B / RT1) * B;
    } else {
        RT1 = HALF * RT;
        RT2 = -HALF * RT;
    }
}

// Main function to test the DLAE2 function
int main() {
    double A, B, C, RT1, RT2;

    // Test case 1
    A = 4.0;
    B = 1.0;
    C = 3.0;
    DLAE2(A, B, C, RT1, RT2);
    std::cout << "Test case 1: Eigenvalues should be 5 and 2." << std::endl;
    std::cout << "Computed eigenvalues: " << RT1 << " " << RT2 << std::endl;

    // Add more test cases as needed, varying A, B, and C to test different scenarios

    return 0;
}

File: data/readable_format/fortran/file054.f90
! dlae2.f90
SUBROUTINE DLAE2( A, B, C, RT1, RT2 )
      DOUBLE PRECISION   A, B, C, RT1, RT2
      DOUBLE PRECISION   ONE
      PARAMETER          ( ONE = 1.0D0 )
      DOUBLE PRECISION   TWO
      PARAMETER          ( TWO = 2.0D0 )
      DOUBLE PRECISION   ZERO
      PARAMETER          ( ZERO = 0.0D0 )
      DOUBLE PRECISION   HALF
      PARAMETER          ( HALF = 0.5D0 )
      DOUBLE PRECISION   AB, ACMN, ACMX, ADF, DF, RT, SM, TB
      INTRINSIC          ABS, SQRT
      SM = A + C
      DF = A - C
      ADF = ABS( DF )
      TB = B + B
      AB = ABS( TB )
      IF( ABS( A ).GT.ABS( C ) ) THEN
         ACMX = A
         ACMN = C
      ELSE
         ACMX = C
         ACMN = A
      END IF
      IF( ADF.GT.AB ) THEN
         RT = ADF*SQRT( ONE+( AB / ADF )**2 )
      ELSE IF( ADF.LT.AB ) THEN
         RT = AB*SQRT( ONE+( ADF / AB )**2 )
      ELSE
         RT = AB*SQRT( TWO )
      END IF
      IF( SM.LT.ZERO ) THEN
         RT1 = HALF*( SM-RT )
         RT2 = ( ACMX / RT1 )*ACMN - ( B / RT1 )*B
      ELSE IF( SM.GT.ZERO ) THEN
         RT1 = HALF*( SM+RT )
         RT2 = ( ACMX / RT1 )*ACMN - ( B / RT1 )*B
      ELSE
         RT1 = HALF*RT
         RT2 = -HALF*RT
      END IF
      RETURN
END

PROGRAM TestDLAE2
  IMPLICIT NONE
  DOUBLE PRECISION :: A, B, C, RT1, RT2

  ! Test case 1
  A = 4.0D0
  B = 1.0D0
  C = 3.0D0
  CALL DLAE2(A, B, C, RT1, RT2)
  PRINT *, 'Test case 1: Eigenvalues should be 5 and 2.'
  PRINT *, 'Computed eigenvalues: ', RT1, RT2

  ! Add more test cases as needed, varying A, B, and C to test different scenarios

END PROGRAM TestDLAE2